src/pool/spawn.rs

```diff
@@ -120,6 +125,61 @@ impl<T> QueueCore<T> {
             }
         }
     }
+
+    /// Scale workers.
+    pub fn scale_workers(&self, new_thread_count: usize) -> bool {
+        if new_thread_count < self.config.min_thread_count
+            || new_thread_count > self.config.max_thread_count
+        {
+            return false;
+        }
+        let current_thread_count = self
+            .config
+            .core_thread_count
+            .swap(new_thread_count, Ordering::SeqCst);
+        let gap = new_thread_count as isize - current_thread_count as isize;
+        if gap > 0 {
+            let addr = self as *const QueueCore<T> as usize | 1;
+            for _ in 0..gap {
+                unsafe {
+                    parking_lot_core::unpark_one(addr, |_| UnparkToken(0));
+                }
+            }
+            self.core_workers.store(new_thread_count, Ordering::SeqCst);
+        } else {
+            let addr = self as *const QueueCore<T> as usize;
+            unsafe {
+                parking_lot_core::unpark_all(addr, UnparkToken(0));
+            }
+        }
+        true
+    }
+
+    pub fn need_to_park_worker(&self) -> bool {
+        let mut core_workers = self.core_workers.load(Ordering::SeqCst);
+        let core_thread_count = self.config.core_thread_count.load(Ordering::SeqCst);
+        if core_workers <= core_thread_count {
+            return false;
+        }
+        loop {
+            match self.core_workers.compare_exchange_weak(
+                core_workers,
+                core_workers - 1,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                Ok(n) => {
+                    if n > core_thread_count {
```

Comment:
What if `core_thread_count` is updated after the loading at L160?

Reply:
This path will only be reached when in scaling down.
- If the modification is also due to scale down, then it does not matter. The next time this function is executed by some thread, it will be re-judged with latest value, and do scale down.
- If the modification is due to scale up but `self.core_workers` hasn't been updated by `scale_workers`, eg: it's doing `unpack_one`, then it doesn't matter too.
-  If the modification is due to scale up but `self.core_workers` has been updated by `scale_workers` too and before `self.core_workers.compare_exchange_weak`,  the value of `core_workers` may be incorrectly reduced, re-execute this scale up operation can be repaired.

I think the probability of triggering this problem in actual scenes is quite small, because scale itself should be a very low frequency event and also needs to be treated with caution.

Reply:
> actual scenes is quite small

Can it be made correct?

Reply:
I did not think of a solution before, because two atomic variables have to be manipulated

Reply:
How about change `core_workers` by each thread at reentrant_start/end, thus its count can always represent the exact number of active threads.